### PR TITLE
GH-24826: [Java] Add DUV.setOffset method

### PR DESCRIPTION
**Pull request by jarohen**
### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes - the addition of a public DUV.setOffset method
* GitHub Issue: #24826

### CODE REVIEW

File: java/vector/src/main/codegen/templates/DenseUnionVector.java [hunk 1]
```diff
@@ -908,6 +908,14 @@ private int getTypeBufferValueCapacity() {
     return (int) typeBuffer.capacity() / TYPE_WIDTH;
   }
 
+  public void setOffset(int index, int offset) {
+    while (index >= getOffsetBufferValueCapacity()) {
```
Comment (vibhatha) [inline java/vector/src/main/codegen/templates/DenseUnionVector.java:912]: just curious: when to reallocate?

File: java/vector/src/test/java/org/apache/arrow/vector/TestDenseUnionVector.java [hunk 2]
```diff
@@ -99,6 +99,44 @@ public void testDenseUnionVector() throws Exception {
...
   }
 
+  @Test
+  public void testSetOffset() {
+    try (DenseUnionVector duv = DenseUnionVector.empty("foo", allocator)) {
+      duv.allocateNew();
+      byte i32TypeId = duv.registerNewTypeId(Field.notNullable("i32", MinorType.INT.getType()));
+      byte f64TypeId = duv.registerNewTypeId(Field.notNullable("f64", MinorType.FLOAT8.getType()));
+
+      IntVector i32Vector = ((IntVector) duv.addVector(i32TypeId, new IntVector("i32", allocator)));
```
Comment (vibhatha) [inline java/vector/src/test/java/org/apache/arrow/vector/TestDenseUnionVector.java:109]: Should we use try-with-resources?

**Review by vibhatha — COMMENTED** [review]
Thanks for working on this.
Left a comment. Need to checkout and try a few ideas.

File: java/vector/src/test/java/org/apache/arrow/vector/TestDenseUnionVector.java (hunk 2 above)
Comment (lidavidm) [reply java/vector/src/test/java/org/apache/arrow/vector/TestDenseUnionVector.java:109]: No, the union vector will close it itself.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**